rpc/client.go

```diff
@@ -91,219 +95,230 @@ type Client struct {
+	address := addr.String()
+
+	key := network + address
+
 	if !context.DisableCache {
-		if c, ok := clients[addr.String()]; ok {
-			clientMu.Unlock()
-			return c
+		if c, ok := clients[key]; ok {
+			return c, nil
 		}
 	}
+
+	tlsConfig, err := context.GetClientTLSConfig()
+	if err != nil {
+		return nil, err
+	}
+
 	c := &Client{
-		addr:         addr,
-		Ready:        make(chan struct{}),
-		Closed:       make(chan struct{}),
-		clock:        context.localClock,
-		remoteClocks: context.RemoteClocks,
-		cached:       !context.DisableCache,
+		Closed:        make(chan struct{}),
+		key:           key,
+		network:       network,
+		remoteAddress: address,
+		tlsConfig:     tlsConfig,
+		clock:         context.localClock,
+		remoteClocks:  context.RemoteClocks,
 	}
-	if !context.DisableCache {
-		clients[c.Addr().String()] = c
+
+	c.healthy.Store(make(chan struct{}))
+
+	if err = c.connect(); err != nil {
+		return nil, err
 	}
-	clientMu.Unlock()
+
+	clients[key] = c
+
+	retryOpts := clientRetryOptions
+	retryOpts.Stopper = context.Stopper
 
 	context.Stopper.RunWorker(func() {
-		if err := c.connect(opts, context); err != nil {
-			log.Errorf("client %s failed to connect: %v", c.addr, err)
-		} else {
-			// Signal client is ready by closing Ready channel.
-			log.Infof("client %s connected", c.addr)
-			close(c.Ready)
+		<-context.Stopper.ShouldStop()
 
-			// Launch periodic heartbeat. This blocks until the client is to be closed.
-			c.runHeartbeat(context.Stopper)
-		}
 		c.Close()
 	})
-	return c
+
+	go func() {
+		c.runHeartbeat(retryOpts)
+		c.rpcClient.Close()
+	}()
+
+	return c, nil
 }
 
-// connect dials the connection in a backoff/retry loop.
-func (c *Client) connect(opts *retry.Options, context *Context) error {
-	// Attempt to dial connection.
-	retryOpts := clientRetryOptions
-	if opts != nil {
-		retryOpts = *opts
+// Go wraps net/rpc.Client.Go and provides additional cockroach-specific error
+// handling.
+func (c *Client) Go(serviceMethod string, args interface{}, reply interface{}, done chan *rpc.Call) *rpc.Call {
+	call := c.rpcClient.Go(serviceMethod, args, reply, nil)
+	newCall := &rpc.Call{}
+	if done == nil {
+		newCall.Done = make(chan *rpc.Call, 1)
+	} else {
+		newCall.Done = done
 	}
-	retryOpts.Stopper = context.Stopper
 
-	for r := retry.Start(retryOpts); r.Next(); {
-		tlsConfig, err := context.GetClientTLSConfig()
-		if err != nil {
-			// Problem loading the TLS config. Retrying will not help.
-			return err
-		}
+	go func() {
+		finishedCall := <-call.Done
+		newCall.ServiceMethod = finishedCall.ServiceMethod
+		newCall.Args = finishedCall.Args
+		newCall.Reply = finishedCall.Reply
+		newCall.Error = finishedCall.Error
 
-		conn, err := tlsDialHTTP(c.addr.Network(), c.addr.String(), tlsConfig)
-		if err != nil {
-			// Retry if the error is temporary, otherwise fail fast.
-			if t, ok := err.(net.Error); ok && t.Temporary() {
-				if log.V(1) {
-					log.Warning(err)
+		if newCall.Error != nil {
```

Comment:
Sounds good to me. One problem with this is that the "client connection died" error is retryable _only_ if it comes from this special sauce client that does automatic reconnection, so I don't think it'll help us avoid the extra channel and goroutine in this case.
However, I can probably implement this using `client.Call` rather than `client.Go` which will save one goroutine. I'll do that now and reopen this.